crates/adapters/src/transport/s3.rs

```diff
@@ -538,30 +540,50 @@ impl S3InputReader {
                                 continue;
                             };
 
-                            let result = client
-                                .get_object(
-                                    &bucket_name,
-                                    &partially_processed_key.key,
-                                    start_offset,
-                                    None,
-                                )
-                                .await;
-                            let mut object = match result {
-                                Ok(ret) => ret,
-                                Err(e) => {
-                                    consumer.error(
-                                        false,
-                                        anyhow!(
-                                            "could not fetch object '{}': {e:?}",
-                                            &partially_processed_key.key
-                                        ),
-                                        Some("s3-obj-fetch"),
-                                    );
-                                    // Mark key as fully processed.
-                                    partially_processed_keys.update(key_index, None).await;
-                                    continue;
+                            // Retry fetching the object for `max_retries`, with exponential
+                            // backoff.
+                            // If we still fail after that, return the error.
+                            async fn retry_fetching_object(
+                                client: Arc<dyn S3Api>,
+                                bucket_name: &str,
+                                key: &str,
+                                start_offset: u64,
+                                end: Option<u64>,
+                                max_retries: u32,
+                            ) -> Result<GetObjectOutput, anyhow::Error>
+                            {
+                                let mut retry_counter = 0;
+                                let backoff =
+                                    |count: u32| Duration::from_millis(100 * 2_u64.pow(count));
```

Comment:
Let's add an upper bound of maybe 20 seconds or something like that. Also, please add some info-level logging that kicks in after backing off for N seconds.